apps/passport-server/src/services/generic-issuance/pipelines/PretixPipeline.ts

```diff
@@ -7,37 +52,723 @@ import {
+
+    return ticketPCD;
+  }
+
+  private async checkinPretixTicketPCDs(
+    request: GenericIssuanceCheckInRequest
+  ): Promise<CheckTicketInResponseValue> {
+    logger(
+      LOG_TAG,
+      `got request to check in tickets with request ${JSON.stringify(request)}`
+    );
+
+    const signaturePCD = await SemaphoreSignaturePCDPackage.deserialize(
+      request.credential.pcd
+    );
+    const signaturePCDValid =
+      await SemaphoreSignaturePCDPackage.verify(signaturePCD);
+
+    if (!signaturePCDValid) {
+      throw new Error("credential signature invalid");
+    }
+
+    const payload: GenericCheckinCredentialPayload = JSON.parse(
+      signaturePCD.claim.signedMessage
+    );
+
+    const checkerEmailPCD = await EmailPCDPackage.deserialize(
+      payload.emailPCD.pcd
+    );
+
+    const checkerTickets = await this.db.loadByEmail(
+      this.id,
+      checkerEmailPCD.claim.emailAddress
+    );
+
+    const ticketToCheckIn = await EdDSATicketPCDPackage.deserialize(
+      payload.ticketToCheckIn.pcd
+    );
+
+    const pretixEventId = this.eddsaTicketToPretixEventId(ticketToCheckIn);
+
+    const pretixProductId = this.eddsaTicketToPretixProductId(ticketToCheckIn);
+
+    const eventConfig = this.definition.options.events.find(
+      (e) => e.externalId === pretixEventId
+    );
+
+    if (!eventConfig) {
+      throw new Error(
+        `${pretixEventId} has no corresponding event configuration`
+      );
+    }
+
+    const productConfig = eventConfig.products.find(
+      (t) => t.externalId === pretixProductId
+    );
+
+    if (!productConfig) {
+      throw new Error(
+        `${pretixProductId} has no corresponding product configuration`
+      );
+    }
+
+    const checkerEventTickets = checkerTickets.filter(
+      (t) => t.eventId === eventConfig.genericIssuanceId
+    );
+    const checkerProducts = checkerEventTickets.map((t) => {
+      const productConfig = eventConfig.products.find(
+        (p) => p.genericIssuanceId === t.productId
+      );
+      return productConfig;
+    });
+    const hasSuperUserProductTicket = checkerProducts.find(
+      (t) => t?.isSuperUser
+    );
+
+    if (!hasSuperUserProductTicket) {
+      throw new Error(
+        `user ${checkerEmailPCD.claim.emailAddress} doesn't have a superuser ticket`
+      );
+    }
+
+    const ticketAtom = await this.db.loadById(
+      this.id,
+      ticketToCheckIn.claim.ticket.ticketId
+    );
+    if (!ticketAtom) {
+      throw new Error(
+        `Could not load atom for ticket id ${ticketToCheckIn.claim.ticket.ticketId}`
+      );
+    }
+
+    const checkinLists = await this.api.fetchEventCheckinLists(
+      this.definition.options.pretixOrgUrl,
+      this.definition.options.pretixAPIKey,
+      pretixEventId
+    );
+
+    if (checkinLists.length === 0) {
+      throw new Error(
+        `Could not fetch check-in lists for ${eventConfig.genericIssuanceId}`
+      );
+    }
+
+    // TODO: error handling
+    await this.api.pushCheckin(
```

Comment:
just checking but this does throw on the pretix end if the ticket is already checked in right? i think based on the test you wrote it does, but i wanted to make sure. it's important we don't make it possible to check in a ticket twice with zupass.

Reply:
It does, yes